src/nvim/optionstr.c

```diff
@@ -1261,320 +1798,78 @@ char *did_set_string_option(int opt_idx, char **varp, char *oldval, char *errbuf
              || gvarp == &p_inex
              || gvarp == &p_inde
              || varp == &p_pex) {  // '*expr' options
-    char **p_opt = NULL;
-
-    // If the option value starts with <SID> or s:, then replace that with
-    // the script identifier.
-
-    if (varp == &p_dex) {  // 'diffexpr'
-      p_opt = &p_dex;
-    }
-    if (varp == &curwin->w_p_fde) {  // 'foldexpr'
-      p_opt = &curwin->w_p_fde;
-    }
-    if (varp == &curwin->w_p_fdt) {  // 'foldtext'
-      p_opt = &curwin->w_p_fdt;
-    }
-    if (gvarp == &p_fex) {  // 'formatexpr'
-      p_opt = &curbuf->b_p_fex;
-    }
-    if (gvarp == &p_inex) {  // 'includeexpr'
-      p_opt = &curbuf->b_p_inex;
-    }
-    if (gvarp == &p_inde) {  // 'indentexpr'
-      p_opt = &curbuf->b_p_inde;
-    }
-    if (varp == &p_pex) {  // 'patchexpr'
-      p_opt = &p_pex;
-    }
-
-    if (p_opt != NULL) {
-      char *name = get_scriptlocal_funcname(*p_opt);
-      if (name != NULL) {
-        free_string_option(*p_opt);
-        *p_opt = name;
-      }
-    }
-
-    if (varp == &curwin->w_p_fde && foldmethodIsExpr(curwin)) {
-      foldUpdateAll(curwin);
-    }
+    did_set_optexpr(curbuf, curwin, varp, gvarp);
   } else if (gvarp == &p_cfu) {  // 'completefunc'
-    if (set_completefunc_option() == FAIL) {
-      errmsg = e_invarg;
-    }
+    set_completefunc_option(&errmsg);
   } else if (gvarp == &p_ofu) {  // 'omnifunc'
-    if (set_omnifunc_option() == FAIL) {
-      errmsg = e_invarg;
-    }
+    set_omnifunc_option(&errmsg);
   } else if (gvarp == &p_tsrfu) {  // 'thesaurusfunc'
-    if (set_thesaurusfunc_option() == FAIL) {
-      errmsg = e_invarg;
-    }
+    set_thesaurusfunc_option(&errmsg);
   } else if (varp == &p_opfunc) {  // 'operatorfunc'
-    if (set_operatorfunc_option() == FAIL) {
-      errmsg = e_invarg;
-    }
+    set_operatorfunc_option(&errmsg);
   } else if (varp == &p_qftf) {  // 'quickfixtextfunc'
-    if (qf_process_qftf_option() == FAIL) {
-      errmsg = e_invarg;
-    }
+    qf_process_qftf_option(&errmsg);
   } else if (gvarp == &p_tfu) {  // 'tagfunc'
-    if (set_tagfunc_option() == FAIL) {
-      errmsg = e_invarg;
-    }
+    set_tagfunc_option(&errmsg);
   } else {
-    // Options that are a list of flags.
-    p = NULL;
-    if (varp == &p_ww) {  // 'whichwrap'
-      p = WW_ALL;
-    }
-    if (varp == &p_shm) {  // 'shortmess'
-      p = SHM_ALL;
-    } else if (varp == &(p_cpo)) {  // 'cpoptions'
-      p = CPO_VI;
-    } else if (varp == &(curbuf->b_p_fo)) {  // 'formatoptions'
-      p = FO_ALL;
-    } else if (varp == &curwin->w_p_cocu) {  // 'concealcursor'
-      p = COCU_ALL;
-    } else if (varp == &p_mouse) {  // 'mouse'
-      p = MOUSE_ALL;
-    }
-    if (p != NULL) {
-      for (s = *varp; *s; s++) {
-        if (vim_strchr(p, (uint8_t)(*s)) == NULL) {
-          errmsg = illegal_char(errbuf, errbuflen, *s);
-          break;
-        }
-      }
-    }
+    did_set_option_listflags(varp, errbuf, errbuflen, &errmsg);
```

Comment:
sometimes you add  curwin/curbuf parameters to the new function, sometimes not (as here). Is there any reasoning behind that or it's just as you go and there will be followup PRs ?

Reply:
The sole reason is that I didn't notice the dependency on curbuf/curwin. Most of these functions are low-level, so this is the place to start removing the dependency on curbuf/curwin.